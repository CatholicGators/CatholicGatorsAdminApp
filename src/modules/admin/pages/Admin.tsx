import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Theme,
  createStyles,
  withStyles,
  Paper,
  Table,
  TableHead,
  TableRow,
  TableCell,
  TableBody,
  Avatar,
  Switch,
  Toolbar,
  Typography,
  CircularProgress
} from '@material-ui/core';

import {
  getUsers,
  updateUser
} from '../../../redux/actions/admin/adminActions';

const styles = (theme: Theme) => createStyles({
  tableWrapper: {
    maxWidth: '1000px',
    marginTop: '50px',
    marginLeft: 'auto',
    marginRight: 'auto'
  },
  tableCard: {
    overflowX: 'auto'
  },
  tableLoadingContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: '300px',
  }
})

type Props = {
  classes: any
  users: any
  getUsers: () => any,
  updateUser: (user) => any
}

export class Admin extends Component<Props> {
  componentDidMount() {
    this.props.getUsers()
  }

  handleApprovedToggle(user, checked) {
    this.props.updateUser({
      ...user,
      data: {
        ...user.data,
        isApproved: checked
      }
    })
  }

  handleAdminToggle(user, checked) {
    this.props.updateUser({
      ...user,
      data: {
        ...user.data,
        isAdmin: checked
      }
    })
  }

  render() {
    const { classes, users } = this.props;

    return (
      <div className={classes ? classes.tableWrapper : null}>
        <Paper className={classes ? classes.tableCard : null}>
          <Toolbar>
            <Typography variant="h6">
              Users
            </Typography>
          </Toolbar>
          <Table>
            <TableHead>
              <TableRow>
                <TableCell>Profile Pic</TableCell>
                <TableCell>Name</TableCell>
                <TableCell>Email</TableCell>
                <TableCell>Approved</TableCell>
                <TableCell>Admin</TableCell>
              </TableRow>
            </TableHead>
            {users ? 
              <TableBody>
                {users.map(user => (
                  <TableRow key={user.id}>
                    <TableCell>
                      <Avatar src={user.photoURL} />
                    </TableCell>
                    <TableCell>{user.name}</TableCell>
                    <TableCell>{user.email}</TableCell>
                    <TableCell>
                      <Switch
                        checked={user.isApproved}
                        onChange={(_, checked) => this.handleApprovedToggle(user, checked)}/>
                    </TableCell>
                    <TableCell>
                      <Switch
                        disabled={!user.isApproved}
                        checked={user.isAdmin}
                        onChange={(_, checked) => this.handleAdminToggle(user, checked)}
                      />
                    </TableCell>
                  </TableRow>
                ))}
              </TableBody>
            : null }
          </Table>
          {!users ?
            <div className={classes ? classes.tableLoadingContainer : null}>
              <CircularProgress />
            </div>
          : null}
        </Paper>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  users: state.admin.users
})

const mapDispatchToProps = dispatch => ({
  getUsers: () => dispatch(getUsers()),
  updateUser: user => dispatch(updateUser(user))
})

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Admin))
