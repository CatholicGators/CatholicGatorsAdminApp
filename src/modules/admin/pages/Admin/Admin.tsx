import React, { Component } from "react"
import { Route, Redirect, NavLink } from "react-router-dom"
import {
    Theme,
    createStyles,
    withStyles,
    ListItem,
    ListItemText,
    List,
    ListItemIcon,
} from "@material-ui/core"
import { VpnKey, Group } from "@material-ui/icons"
import UserTable from "../../modules/users/components/UserTable/UserTable"
import Interests from "../../modules/interests/components/Interests/Interests"

export const styles = (theme: Theme) =>
    createStyles({
        pageWrapper: {
            display: "flex",
            margin: "50px 0px",
            padding: `0px ${theme.spacing(1)}px`,
            width: "100%",
        },
        pageNav: {
            width: 250,
            minWidth: 250,
            padding: 0,
            marginRight: `${theme.spacing(1)}px`,
        },
        navLink: {
            textDecoration: "none",
            "&:visited": {
                color: "inherit",
            },
        },
        colorInitial: {
            color: "initial",
        },
    })

type Props = {
    classes: any
    match: any
    location: any
}

export type MenuLink = {
    text: string
    href: string
    icon: any
}
export const menuLinks: Array<MenuLink> = [
    {
        text: "Users",
        href: "/users",
        icon: Group,
    },
    {
        text: "Interests",
        href: "/interests",
        icon: VpnKey,
    },
]

export class Admin extends Component<Props> {
    render() {
        const { classes, match, location } = this.props

        return (
            <div className={classes.pageWrapper}>
                <List className={classes.pageNav}>
                    {menuLinks.map((link) => (
                        <NavLink
                            exact
                            key={link.text}
                            to={`${match.url}${link.href}`}
                            className={classes.navLink}
                        >
                            <ListItem
                                button
                                className={classes.colorInitial}
                                selected={
                                    location.pathname ===
                                    `${match.url}${link.href}`
                                }
                            >
                                <ListItemIcon>
                                    <link.icon />
                                </ListItemIcon>
                                <ListItemText primary={link.text} />
                            </ListItem>
                        </NavLink>
                    ))}
                </List>
                <Route
                    exact
                    path={`${match.url}`}
                    render={() => <Redirect to={`${match.url}/users`} />}
                />
                <Route path={`${match.url}/users`} component={UserTable} />
                <Route path={`${match.url}/interests`} component={Interests} />
            </div>
        )
    }
}

export default withStyles(styles)(Admin)
