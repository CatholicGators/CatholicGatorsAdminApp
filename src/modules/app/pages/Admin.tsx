import React, { Component } from 'react';
import { listenForUser } from '../../../redux/actions/auth/authActions';
import { connect } from 'react-redux';

type Props = {
  listenForUser: () => any;
}

export class Admin extends Component<Props> {
  render() {
    return (
      <div>
        <h1>Admin!</h1>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  listenForUser: () => dispatch(listenForUser())
});

export default connect(null, mapDispatchToProps)(Admin);
