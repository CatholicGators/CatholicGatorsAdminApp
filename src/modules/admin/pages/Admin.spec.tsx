import React from 'react'
import { shallow } from 'enzyme';

import { Admin } from './Admin'
import { TableHead, Checkbox } from '@material-ui/core';

describe('Admin', () => {
    let props, wrapper

    beforeEach(() => {
        props = {
            getUsers: jest.fn(),
            users: [
                {
                    id: "1",
                    name: "Joey",
                    isAdmin: false
                },
                {
                    id: "2",
                    name: "Ryan",
                    isAdmin: false
                },
                {
                    id: "3",
                    name: "Espresso Shot Capuccino",
                    isAdmin: true
                }
            ]
        }
        wrapper = shallow(<Admin {...props} />)
    })

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    it('should call getUsers when mounted', () => {
        expect(props.getUsers).toHaveBeenCalled()
    })

    it('shows the circular progress when run loading the users', () => {
        wrapper.setProps({ users: undefined })
        expect(wrapper.find('#loading-spinner').exists()).toBe(true)
    })

    it('sets the selected array to contain all of the ids when the table head checkbox is clicked', () => {
        const header = wrapper.find(TableHead)
        const checkbox = header.find(Checkbox)

        checkbox.simulate('change', { target: { checked: true } })

        const selected = wrapper.state().selected
        expect(selected.length).toEqual(props.users.length)
        for(var i = 0; i < props.users.length; i++) {
            expect(selected).toContain(props.users[i].id)
        }
    })
})
