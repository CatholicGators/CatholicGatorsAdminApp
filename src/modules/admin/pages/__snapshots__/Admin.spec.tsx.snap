// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Admin
    getUsers={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <WithStyles(Paper)
        className={null}
      >
        <WithStyles(Toolbar)>
          <WithStyles(Typography)
            variant="h6"
          >
            Users
          </WithStyles(Typography)>
        </WithStyles(Toolbar)>
        <WithStyles(Table)>
          <WithStyles(TableHead)>
            <WithStyles(TableRow)>
              <WithStyles(TableCell)>
                Profile Pic
              </WithStyles(TableCell)>
              <WithStyles(TableCell)>
                Name
              </WithStyles(TableCell)>
              <WithStyles(TableCell)>
                Email
              </WithStyles(TableCell)>
              <WithStyles(TableCell)>
                Approved
              </WithStyles(TableCell)>
              <WithStyles(TableCell)>
                Admin
              </WithStyles(TableCell)>
            </WithStyles(TableRow)>
          </WithStyles(TableHead)>
          <WithStyles(TableBody) />
        </WithStyles(Table)>
      </WithStyles(Paper)>,
      "className": null,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Toolbar)>
            <WithStyles(Typography)
              variant="h6"
            >
              Users
            </WithStyles(Typography)>
          </WithStyles(Toolbar)>,
          <WithStyles(Table)>
            <WithStyles(TableHead)>
              <WithStyles(TableRow)>
                <WithStyles(TableCell)>
                  Profile Pic
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Name
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Email
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Approved
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Admin
                </WithStyles(TableCell)>
              </WithStyles(TableRow)>
            </WithStyles(TableHead)>
            <WithStyles(TableBody) />
          </WithStyles(Table)>,
        ],
        "className": null,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(Typography)
              variant="h6"
            >
              Users
            </WithStyles(Typography)>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Users",
              "variant": "h6",
            },
            "ref": null,
            "rendered": "Users",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <WithStyles(TableHead)>
                <WithStyles(TableRow)>
                  <WithStyles(TableCell)>
                    Profile Pic
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Name
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Email
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Approved
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Admin
                  </WithStyles(TableCell)>
                </WithStyles(TableRow)>
              </WithStyles(TableHead)>,
              <WithStyles(TableBody) />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <WithStyles(TableRow)>
                  <WithStyles(TableCell)>
                    Profile Pic
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Name
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Email
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Approved
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Admin
                  </WithStyles(TableCell)>
                </WithStyles(TableRow)>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    <WithStyles(TableCell)>
                      Profile Pic
                    </WithStyles(TableCell)>,
                    <WithStyles(TableCell)>
                      Name
                    </WithStyles(TableCell)>,
                    <WithStyles(TableCell)>
                      Email
                    </WithStyles(TableCell)>,
                    <WithStyles(TableCell)>
                      Approved
                    </WithStyles(TableCell)>,
                    <WithStyles(TableCell)>
                      Admin
                    </WithStyles(TableCell)>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Profile Pic",
                    },
                    "ref": null,
                    "rendered": "Profile Pic",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Name",
                    },
                    "ref": null,
                    "rendered": "Name",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Email",
                    },
                    "ref": null,
                    "rendered": "Email",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Approved",
                    },
                    "ref": null,
                    "rendered": "Approved",
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": "Admin",
                    },
                    "ref": null,
                    "rendered": "Admin",
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": null,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <WithStyles(Paper)
          className={null}
        >
          <WithStyles(Toolbar)>
            <WithStyles(Typography)
              variant="h6"
            >
              Users
            </WithStyles(Typography)>
          </WithStyles(Toolbar)>
          <WithStyles(Table)>
            <WithStyles(TableHead)>
              <WithStyles(TableRow)>
                <WithStyles(TableCell)>
                  Profile Pic
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Name
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Email
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Approved
                </WithStyles(TableCell)>
                <WithStyles(TableCell)>
                  Admin
                </WithStyles(TableCell)>
              </WithStyles(TableRow)>
            </WithStyles(TableHead)>
            <WithStyles(TableBody) />
          </WithStyles(Table)>
        </WithStyles(Paper)>,
        "className": null,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(Toolbar)>
              <WithStyles(Typography)
                variant="h6"
              >
                Users
              </WithStyles(Typography)>
            </WithStyles(Toolbar)>,
            <WithStyles(Table)>
              <WithStyles(TableHead)>
                <WithStyles(TableRow)>
                  <WithStyles(TableCell)>
                    Profile Pic
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Name
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Email
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Approved
                  </WithStyles(TableCell)>
                  <WithStyles(TableCell)>
                    Admin
                  </WithStyles(TableCell)>
                </WithStyles(TableRow)>
              </WithStyles(TableHead)>
              <WithStyles(TableBody) />
            </WithStyles(Table)>,
          ],
          "className": null,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <WithStyles(Typography)
                variant="h6"
              >
                Users
              </WithStyles(Typography)>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Users",
                "variant": "h6",
              },
              "ref": null,
              "rendered": "Users",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <WithStyles(TableHead)>
                  <WithStyles(TableRow)>
                    <WithStyles(TableCell)>
                      Profile Pic
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Name
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Email
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Approved
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Admin
                    </WithStyles(TableCell)>
                  </WithStyles(TableRow)>
                </WithStyles(TableHead)>,
                <WithStyles(TableBody) />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <WithStyles(TableRow)>
                    <WithStyles(TableCell)>
                      Profile Pic
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Name
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Email
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Approved
                    </WithStyles(TableCell)>
                    <WithStyles(TableCell)>
                      Admin
                    </WithStyles(TableCell)>
                  </WithStyles(TableRow)>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": Array [
                      <WithStyles(TableCell)>
                        Profile Pic
                      </WithStyles(TableCell)>,
                      <WithStyles(TableCell)>
                        Name
                      </WithStyles(TableCell)>,
                      <WithStyles(TableCell)>
                        Email
                      </WithStyles(TableCell)>,
                      <WithStyles(TableCell)>
                        Approved
                      </WithStyles(TableCell)>,
                      <WithStyles(TableCell)>
                        Admin
                      </WithStyles(TableCell)>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": "Profile Pic",
                      },
                      "ref": null,
                      "rendered": "Profile Pic",
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": "Name",
                      },
                      "ref": null,
                      "rendered": "Name",
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": "Email",
                      },
                      "ref": null,
                      "rendered": "Email",
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": "Approved",
                      },
                      "ref": null,
                      "rendered": "Approved",
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": "Admin",
                      },
                      "ref": null,
                      "rendered": "Admin",
                      "type": [Function],
                    },
                  ],
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": null,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
